siobrultech_protocols/gem/protocol.py

```diff
@@ -15,16 +16,61 @@
 LOG = logging.getLogger(__name__)
 
 PACKET_HEADER = bytes.fromhex("feff")
+API_RESPONSE_WAIT_TIME_SECONDS = 3.0  # Time to wait for an API response
+PACKET_DELAY_CLEAR_TIME_SECONDS = 3.0  # Time to wait after a packet delay request so that GEM can finish sending any pending packets
 
 
 class PacketProtocol(asyncio.Protocol):
-    def __init__(self, queue: asyncio.Queue):
+    def __init__(
+        self,
+        queue: asyncio.Queue,
+        on_connection_made: Optional[Callable[[GemApi], None]] = None,
+    ):
         self._buffer = bytearray()
         self._queue = queue
-        self._transport: Optional[asyncio.BaseTransport] = None
-
-    def connection_made(self, transport: asyncio.BaseTransport):
+        self._transport: Optional[asyncio.WriteTransport] = None
+        self._api_lock = asyncio.Lock()
+        self._api_mode = False
+        self._on_connection_made = on_connection_made
+
+    async def _send_api_command(self, command: str):
+        async with self._api_lock:  # One API call at a time, please
+            if not self._transport:
+                raise EOFError
+
+            # We're about to send a request on the same channel that the GEM is using to
+            # push data packets to us. To minimize confusion, we ask the GEM to delay packets
+            # for 15 seconds and give it a few seconds to finish sending any in-progress
+            # packets before sending our request.
+            LOG.debug("Requesting packet delay...")
+            self._transport.write(
+                DELAY_NEXT_PACKET.encode()
+            )  # Delay packets for 15 seconds
+            await asyncio.sleep(PACKET_DELAY_CLEAR_TIME_SECONDS)
```

Comment:
I wonder if we can be smarter here and get away with using [`asyncio.Event`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event) to know when we have completed a full packet.  Not required for this PR, but worth thinking about.

Reply:
It's an edge case, since it would only matter for the rare case where we happen to send the delay request while there's a packet in flight. In the common case where our delay request comes in while the GEM is idle, we'd hit whatever timeout we specified on the event. Might as well just wait for the full time all the time.